File: src/utils/xp_calculator.py
"""
XP Calculator - Utility for calculating XP rewards and requirements
"""

from ..models.quest import QuestDifficulty, QuestType


class XPCalculator:
    """Utility class for XP-related calculations"""
    
    # Base XP values by difficulty
    BASE_XP = {
        QuestDifficulty.EASY: 25,
        QuestDifficulty.MEDIUM: 50,
        QuestDifficulty.HARD: 100,
        QuestDifficulty.EPIC: 200,
    }
    
    # Type multipliers
    TYPE_MULTIPLIERS = {
        QuestType.BRAINSTORM: 1.0,
        QuestType.BUILD: 1.2,
        QuestType.PROMOTE: 0.9,
        QuestType.DEBUG: 1.3,
        QuestType.LEARN: 1.1,
        QuestType.REVIEW: 0.8,
        QuestType.DOCUMENTATION: 0.9,
    }
    
    @staticmethod
    def calculate_quest_xp(difficulty: QuestDifficulty, quest_type: QuestType,
                          bonus_multiplier: float = 1.0) -> int:
        """
        Calculate XP reward for a quest.
        
        Args:
            difficulty: Quest difficulty
            quest_type: Quest type
            bonus_multiplier: Optional bonus multiplier (e.g., 1.5 for urgent quests)
            
        Returns:
            Calculated XP value
        """
        base = XPCalculator.BASE_XP.get(difficulty, 50)
        type_mult = XPCalculator.TYPE_MULTIPLIERS.get(quest_type, 1.0)
        
        return int(base * type_mult * bonus_multiplier)
    
    @staticmethod
    def calculate_level_from_xp(xp: int) -> int:
        """
        Calculate level from total XP.
        Uses square root formula: Level = floor(sqrt(xp / 100))
        
        Args:
            xp: Total XP
            
        Returns:
            Calculated level
        """
        if xp < 0:
            return 0
        return int((xp / 100) ** 0.5)
    
    @staticmethod
    def calculate_xp_for_level(level: int) -> int:
        """
        Calculate total XP needed to reach a specific level.
        
        Args:
            level: Target level
            
        Returns:
            Total XP needed
        """
        if level < 0:
            return 0
        return (level ** 2) * 100
    
    @staticmethod
    def calculate_xp_to_next_level(current_xp: int) -> int:
        """
        Calculate XP needed to reach the next level.
        
        Args:
            current_xp: Current total XP
            
        Returns:
            XP needed for next level
        """
        current_level = XPCalculator.calculate_level_from_xp(current_xp)
        next_level_xp = XPCalculator.calculate_xp_for_level(current_level + 1)
        return next_level_xp - current_xp
    
    @staticmethod
    def calculate_progress_to_next_level(current_xp: int) -> float:
        """
        Calculate progress percentage to next level.
        
        Args:
            current_xp: Current total XP
            
        Returns:
            Progress percentage (0-100)
        """
        current_level = XPCalculator.calculate_level_from_xp(current_xp)
        current_level_xp = XPCalculator.calculate_xp_for_level(current_level)
        next_level_xp = XPCalculator.calculate_xp_for_level(current_level + 1)
        
        if next_level_xp == current_level_xp:
            return 100.0
        
        xp_into_level = current_xp - current_level_xp
        xp_needed = next_level_xp - current_level_xp
        
        return (xp_into_level / xp_needed) * 100
    
    @staticmethod
    def calculate_team_xp_multiplier(team_size: int) -> float:
        """
        Calculate XP multiplier based on team size.
        Encourages larger teams with slight bonus.
        
        Args:
            team_size: Number of team members
            
        Returns:
            Multiplier value
        """
        if team_size <= 0:
            return 1.0
        
        # Small bonus for larger teams (max 1.2x at 10+ members)
        return min(1.2, 1.0 + (team_size - 1) * 0.02)
    
    @staticmethod
    def calculate_streak_bonus(streak_days: int) -> float:
        """
        Calculate XP bonus multiplier based on consecutive activity streak.
        
        Args:
            streak_days: Number of consecutive days with activity
            
        Returns:
            Bonus multiplier
        """
        if streak_days <= 0:
            return 1.0
        
        # Bonus increases with streak, max 2x at 30 days
        bonus = 1.0 + min(streak_days * 0.033, 1.0)
        return round(bonus, 2)

File: questboard_app/lib/utils/xp_calculator.dart
import 'package:questboard_app/models/quest.dart';
import 'package:questboard_app/models/achievement.dart';
import 'package:questboard_app/models/user.dart';

/// XP Calculator utility for consistent XP calculations across the app
class XPCalculator {
  /// Calculate XP reward for completing a quest
  static int calculateQuestXP(Quest quest) {
    int baseXP = Quest.getXPForDifficulty(quest.difficulty);
    
    // Bonus for completing all steps
    if (quest.steps.isNotEmpty) {
      final allStepsComplete = quest.steps.every((step) => step.isCompleted);
      if (allStepsComplete) {
        baseXP = (baseXP * 1.1).round();
      }
    }
    
    // Bonus for completing before due date
    if (quest.dueDate != null && quest.completedAt != null) {
      if (quest.completedAt!.isBefore(quest.dueDate!)) {
        final daysEarly = quest.dueDate!.difference(quest.completedAt!).inDays;
        if (daysEarly > 0) {
          baseXP = (baseXP * (1 + (daysEarly * 0.05))).round();
        }
      }
    }
    
    return baseXP;
  }
  
  /// Calculate XP required for next user level
  static int xpForNextLevel(int currentLevel) {
    return currentLevel * 100;
  }
  
  /// Calculate total XP required to reach a specific level
  static int totalXPForLevel(int level) {
    int total = 0;
    for (int i = 1; i < level; i++) {
      total += xpForNextLevel(i);
    }
    return total;
  }
  
  /// Calculate current level from total XP
  static int levelFromXP(int totalXP) {
    int level = 1;
    int xpNeeded = 0;
    
    while (xpNeeded <= totalXP) {
      xpNeeded += xpForNextLevel(level);
      if (xpNeeded <= totalXP) {
        level++;
      }
    }
    
    return level;
  }
  
  /// Calculate progress to next level (0.0 to 1.0)
  static double progressToNextLevel(int totalXP) {
    final currentLevel = levelFromXP(totalXP);
    final xpAtCurrentLevel = totalXPForLevel(currentLevel);
    final xpForNext = xpForNextLevel(currentLevel);
    final currentProgress = totalXP - xpAtCurrentLevel;
    
    return (currentProgress / xpForNext).clamp(0.0, 1.0);
  }
  
  /// Calculate team XP bonus (teams get bonus XP)
  static int calculateTeamXPBonus(int baseXP, int teamSize) {
    // 5% bonus per team member (max 50%)
    final bonusMultiplier = (teamSize * 0.05).clamp(0, 0.5);
    return (baseXP * bonusMultiplier).round();
  }
  
  /// Calculate streak bonus XP
  static int calculateStreakBonus(int baseXP, int streakDays) {
    // 2% bonus per day of streak (max 100%)
    final bonusMultiplier = (streakDays * 0.02).clamp(0, 1.0);
    return (baseXP * bonusMultiplier).round();
  }
  
  /// Calculate XP for achieving an achievement
  static int calculateAchievementXP(Achievement achievement) {
    return Achievement.getXPForTier(achievement.tier);
  }
  
  /// Calculate mascot level from experience
  static int mascotLevelFromXP(int experience) {
    return (experience / 50).floor() + 1;
  }
  
  /// Calculate XP needed for mascot's next level
  static int mascotXPForNextLevel(int currentLevel) {
    return currentLevel * 50;
  }
  
  /// Get skill points based on quest type
  static Map<String, int> calculateSkillPoints(Quest quest) {
    Map<String, int> points = {};
    
    switch (quest.type) {
      case QuestType.brainstorm:
        points['creativity'] = 5;
        points['planning'] = 3;
        break;
      case QuestType.build:
        points['coding'] = 8;
        points['problem_solving'] = 5;
        break;
      case QuestType.promote:
        points['communication'] = 7;
        points['marketing'] = 5;
        break;
      case QuestType.research:
        points['analysis'] = 6;
        points['documentation'] = 4;
        break;
      case QuestType.test:
        points['quality_assurance'] = 7;
        points['attention_to_detail'] = 5;
        break;
      case QuestType.review:
        points['collaboration'] = 6;
        points['feedback'] = 5;
        break;
      case QuestType.learn:
        points['learning'] = 8;
        points['adaptation'] = 4;
        break;
    }
    
    // Difficulty multiplier
    final multiplier = _getDifficultyMultiplier(quest.difficulty);
    points.updateAll((key, value) => (value * multiplier).round());
    
    return points;
  }
  
  static double _getDifficultyMultiplier(QuestDifficulty difficulty) {
    switch (difficulty) {
      case QuestDifficulty.easy:
        return 1.0;
      case QuestDifficulty.medium:
        return 1.5;
      case QuestDifficulty.hard:
        return 2.0;
      case QuestDifficulty.epic:
        return 3.0;
    }
  }
  
  /// Calculate leaderboard score (combination of XP, achievements, and activity)
  static int calculateLeaderboardScore(User user, int questsCompleted, int daysActive) {
    int score = user.totalXP;
    score += user.achievementIds.length * 50;
    score += questsCompleted * 25;
    score += daysActive * 10;
    
    return score;
  }
}
